hummingbot/connector/derivative/phemex_perpetual/phemex_perpetual_api_order_book_data_source.py

```diff
@@ -0,0 +1,234 @@
+import asyncio
+import time
+from collections import defaultdict, namedtuple
+from decimal import Decimal
+from typing import TYPE_CHECKING, Any, Dict, List, Optional
+
+import hummingbot.connector.derivative.phemex_perpetual.phemex_perpetual_constants as CONSTANTS
+import hummingbot.connector.derivative.phemex_perpetual.phemex_perpetual_web_utils as web_utils
+from hummingbot.core.data_type.funding_info import FundingInfo, FundingInfoUpdate
+from hummingbot.core.data_type.order_book_message import OrderBookMessage, OrderBookMessageType
+from hummingbot.core.data_type.perpetual_api_order_book_data_source import PerpetualAPIOrderBookDataSource
+from hummingbot.core.utils.async_utils import safe_ensure_future
+from hummingbot.core.web_assistant.connections.data_types import WSJSONRequest
+from hummingbot.core.web_assistant.web_assistants_factory import WebAssistantsFactory
+from hummingbot.core.web_assistant.ws_assistant import WSAssistant
+
+if TYPE_CHECKING:
+    from hummingbot.connector.derivative.phemex_perpetual.phemex_perpetual_derivative import PhemexPerpetualDerivative
+
+
+TradeStructure = namedtuple("Trade", "timestamp side price amount")
+
+
+class PhemexPerpetualAPIOrderBookDataSource(PerpetualAPIOrderBookDataSource):
+    def __init__(
```

Comment:
We have to remember to redefine in the class variable `_logger` (as it was done in the Phemex user stream data source class). If not this class will use the class variable from the super class.

Reply:
done